src/background-monitor.c

```diff
@@ -0,0 +1,164 @@
+
+#define BACKGROUND_MONITOR_BUS_NAME "org.freedesktop.background.Monitor"
+#define BACKGROUND_MONITOR_OBJECT_PATH "/org/freedesktop/background/monitor"
+
+struct _BackgroundMonitor
+{
+  XdpDbusBackgroundMonitorSkeleton parent_instance;
+
+  GDBusConnection *connection;
+};
+
+static void g_initable_iface_init (GInitableIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (BackgroundMonitor,
+                         background_monitor,
+                         XDP_DBUS_BACKGROUND_TYPE_MONITOR_SKELETON,
+                         G_IMPLEMENT_INTERFACE (G_TYPE_INITABLE, g_initable_iface_init))
+
+static gboolean
+request_freedesktop_background_name (BackgroundMonitor  *self,
+                                     GCancellable       *cancellable,
+                                     GError            **error)
+{
+  g_autoptr(GVariant) reply = NULL;
+  GBusNameOwnerFlags flags;
+  guint32 result;
+
+  flags = G_BUS_NAME_OWNER_FLAGS_REPLACE;
+#if GLIB_CHECK_VERSION(2,54,0)
+  flags |= G_BUS_NAME_OWNER_FLAGS_DO_NOT_QUEUE;
+#endif
+
+  reply = g_dbus_connection_call_sync (self->connection,
+                                       "org.freedesktop.DBus",
+                                       "/org/freedesktop/DBus",
+                                       "org.freedesktop.DBus",
+                                       "RequestName",
+                                       g_variant_new ("(su)", BACKGROUND_MONITOR_BUS_NAME, flags),
+                                       G_VARIANT_TYPE ("(u)"),
+                                       0, -1,
+                                       cancellable,
+                                       error);
+
+  if (!reply)
+    return FALSE;
+
+  g_variant_get (reply, "(u)", &result);
+  if (result != 1)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                   "Failed to own background monitor D-Bus name");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+background_monitor_initable_init (GInitable     *initable,
+                                  GCancellable  *cancellable,
+                                  GError       **error)
+{
+  BackgroundMonitor *self = BACKGROUND_MONITOR (initable);
+  g_autofree char *address = NULL;
+
+  address = g_dbus_address_get_for_bus_sync (G_BUS_TYPE_SESSION, cancellable, error);
+  if (!address)
+    return FALSE;
+
+  self->connection = g_object_new (G_TYPE_DBUS_CONNECTION,
+                                   "address", address,
+                                   "flags", G_DBUS_CONNECTION_FLAGS_AUTHENTICATION_CLIENT |
+#if GLIB_CHECK_VERSION(2,74,0)
+                                            G_DBUS_CONNECTION_FLAGS_CROSS_NAMESPACE |
+#endif
+                                            G_DBUS_CONNECTION_FLAGS_MESSAGE_BUS_CONNECTION,
+                                   "exit-on-close", TRUE,
+                                   NULL);
+
+  if (!g_initable_init (G_INITABLE (self->connection), cancellable, error))
+    return FALSE;
```

Comment:
These two calls can be shrunk into a single `g_initable_new()`.

Reply:
Done